Propagate stream cancellation causes

jrudolph opened the pull request:
This PR is focussed on adding APIs and machinery to provide cancellation causes across stages and reactive streams interfaces. The new APIs are (so far) completely optional:

 * A reactive streams subscription can implement `SubscriptionWithCancelException` if it wants to support that feature
 * There's a new `GraphStage.cancel(cause)` method that can be used to provide a cause (otherwise, `SubscriptionWithCancelException.NoMoreElementsNeeded` is used)
 * There's a new `OutHandler.onDownstreamFinished(cause)` method to override if you want access to (or to propagate) the cancellation cause. If not overridden it calls the old method.
 * Similar APIs (SubSink etc.) have already also been adapted.

These changes should be completely transparent and only noticeable to those parties that implement the APIs.

It still needs to be discussed, if providing a cause should be optional or not. As a follow up we should
 * revisit all the places where our own stages implement `onDownstreamFinish` and make sure to use the new version and propagate causes if possible
 * change the behavior of how `cancelStage` will propagate to output, currently it completes neighboring outputs but it should maybe fail them by default instead.

I'll create follow up tickets to those.

Refs #23908

jrudolph committed:
base functionality

jrudolph committed:
some silencing

jrudolph committed:
remove deprecation for now

jrudolph committed:
fix test compilation

jrudolph committed:
Fix subSource / subSink cancellation handling

jrudolph committed:
GraphStage-fix

jrudolph committed:
grab-fix

helena commented on akka-stream/src/main/scala/akka/stream/impl/ReactiveStreamsCompliance.scala:
```diff
@@ -125,13 +126,14 @@ import org.reactivestreams.{ Subscriber, Subscription }
+      case s: SubscriptionWithCancelException ⇒ s.cancel(cause)
+      case s ⇒ s.cancel()
+    } catch {
+      case NonFatal(t) ⇒ throw new SignalThrewException("It is illegal to throw exceptions from cancel(), rule 3.15", t)
```
Surprised `⇒` made it through

jrudolph replied:
The original change was quite old and before the reformatting, seems scalafmt doesn't convert it automatically.

jrudolph replied:
Replaced.

helena reviewed (commented):
Looks good to me, but I am just getting back into Akka streams after a few years.

jrudolph committed:
fix-restart-flow

jrudolph committed:
Fix ambiguity between complete and cancellation (for isAvailable / grab)

jrudolph committed:
rename lastCancellationCause

jrudolph committed:
add mima

jrudolph committed:
fix cancellation cause propagation in OutputBoundary

jrudolph committed:
cleanup

jrudolph committed:
Fix cancellation cause propagation in SubSink

jrudolph committed:
remove warning silencing in Graph

jrudolph committed:
GraphStage cleanup

jrudolph committed:
no unicode arrows

helena commented on akka-stream/src/main/scala/akka/stream/SubscriptionWithCancelException.scala:
```diff
@@ -0,0 +1,24 @@
+object SubscriptionWithCancelException {
+  case object NoMoreElementsNeeded extends RuntimeException with NoStackTrace
+  case object StageWasCompleted extends RuntimeException with NoStackTrace
+}
```
I wonder if `SubscriptionWithCancellationException` is better

helena replied:
Now I see it is better in line with `cancel`, disregard.

jrudolph replied:
Yeah, hard to say. Maybe the default exception should have a less verbose shortcut anyway and then the name doesn't matter as much in user facing API.

chbatey commented on akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala:
```diff
@@ -548,17 +560,24 @@ abstract class GraphStageLogic private[stream] (val inCount: Int, val outCount:
-      connection.slot = Failed(failed.ex, Empty)
-      elem
+
+      if ((connection.portState & (InReady | InFailed)) == (InReady | InFailed)) {
```
a comment for each of these cases for the scenario?

jrudolph replied:
Yep, added here and elsewhere.

chbatey commented on akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala:
```diff
@@ -1772,14 +1804,29 @@ trait OutHandler {
   def onPull(): Unit
 
+  // Hack to make sure that old `onDownstreamFinish` can be called without losing the cause in the default implementation
+  private[this] var _lastCancellationCause: Throwable = _
```
:+1:

chbatey reviewed (commented):
Looks good

raboof commented on akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala:
```diff
@@ -1772,14 +1804,29 @@ trait OutHandler {
-  }
+  // FIXME: we should add this deprecation but first we need to fix all our own usages
+  // @deprecatedOverride("Override method that provides cause.", since = "2.6.0")
+  def onDownstreamFinish(): Unit =
```
AFAICS this change assumes client code may extend `onDownstreamFinish()`, but may never call it (aside from the `super.onDownstreamFinish()` inside the overridden call). That seems valid to me.

Aside: I was actually surprised that this method had an implementation. Most `onDownstreamFinish` implementations don't call `super.onDownstreamFinish()`. Is that a risk?

jrudolph replied:
> AFAICS this change assumes client code may extend `onDownstreamFinish()`, but may never call it (aside from the `super.onDownstreamFinish()` inside the overridden call). That seems valid to me.

That's a good point and many of the bugs I had to fix were related to some infrastructure not calling the right `onDownstreamFinish`. So, we should maybe add a `@deprecate` note with extra info that you shouldn't call it. We can also try to make it `protected[something]` to prevent others from calling it (but not sure if that plays together with still allowing overriding)?

> Aside: I was actually surprised that this method had an implementation. Most onDownstreamFinish implementations don't call super.onDownstreamFinish(). Is that a risk?

You could say that's what all of this is about: being more explicit about what happens in the default case of `onDownstreamFinish` being called. On the other hand, where it is overridden, people already are explicit about the custom behavior they want for their stage (i.e. often ignoring or deferring cancellation).

raboof commented on akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala:
```diff
@@ -1772,14 +1804,29 @@ trait OutHandler {
+  // FIXME: we should add this deprecation but first we need to fix all our own usages
+  // @deprecatedOverride("Override method that provides cause.", since = "2.6.0")
+  def onDownstreamFinish(): Unit =
+    GraphInterpreter.currentInterpreter.activeStage.cancelStage(_lastCancellationCause)
```
Perhaps add a check here that `_lastCancellationCause` is not `null`, to catch people calling `onDownstreamFinished` from user code?

jrudolph replied:
Good suggestion.

jrudolph replied:
Done

raboof commented on akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala:
```diff
@@ -1772,14 +1804,29 @@ trait OutHandler {
+    try {
+      _lastCancellationCause = cause
+      onDownstreamFinish()
+    } finally _lastCancellationCause = null
```
This assumes implementations are not calling `super.onDownstreamFinish()` asynchronously. Sounds safe, but are we sure? Perhaps document it?

jrudolph replied:
Not sure if needed, all of the GraphStage internals are full of mutable state, so if you call any of those methods out of context you are likely to get hurt.

raboof commented on akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala:
```diff
@@ -1505,11 +1505,11 @@ private[stream] object Collect {
+      override def onDownstreamFinish(cause: Throwable): Unit = {
         if (isEnabled(logLevels.onFinish))
-          log.log(logLevels.onFinish, "[{}] Downstream finished.", name)
+          log.log(logLevels.onFinish, "[{}] Downstream finished.", name) // FIXME
```
What needs fixing? Adding the cause to the log?

jrudolph replied:
Yes. Will do.

raboof reviewed (commented):
Some comments, but generally looks pretty convincing. Changes to `portState`/`slot` are a bit hard to fully digest, but the fact that the existing tests still succeed inspires confidence.

raboof reviewed (commented):
Would it be possible to add a spec with a BidiFlow avoiding the fail/cancel race condition by looking at the cancellation cause?

jrudolph commented:
> Would it be possible to add a spec with a BidiFlow avoiding the fail/cancel race condition by looking at the cancellation cause?

That's not yet implemented (though it could now be done more easily). The next step is providing different functionality for `cancelStage` to offer these kinds of behaviors.

jrudolph commented on akka-stream/src/main/scala/akka/stream/SubscriptionWithCancelException.scala:
```diff
@@ -0,0 +1,24 @@
+  def cancel(cause: Throwable): Unit
+}
+object SubscriptionWithCancelException {
+  case object NoMoreElementsNeeded extends RuntimeException with NoStackTrace
```
Not sure if it would happen too often that users would actually have to handle it. New handling code will be in follow-up PRs. I'm not even sure any more if `StageWasCompleted` brings any additional value over `NoMoreElementsNeeded`.

jrudolph commented on akka-stream/src/main/scala/akka/stream/impl/fusing/GraphInterpreter.scala:
```diff
@@ -493,7 +494,9 @@ import akka.stream.snapshot._
       completeConnection(connection.outOwner.stageId)
-      connection.outHandler.onDownstreamFinish()
+      val cause = connection.slot.asInstanceOf[Cancelled].cause
+      connection.slot = Empty
```
Exactly.

jrudolph committed:
Add cancellation cause logging to Flow.log

jrudolph committed:
add more comments about GraphStage portState internals

jrudolph committed:
Add some assertions in onDownstreamFinish to prevent wrong usage

Also deprecate onDownstreamFinish() so that no one calls the wrong one
accidentally

jrudolph committed:
add SubSinkInlet.cancel(cause)

jrudolph committed:
Propagate causes in two other places

jrudolph committed:
improve comment

jrudolph commented on akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala:
```diff
@@ -1772,14 +1808,34 @@ trait OutHandler {
    */
   @throws(classOf[Exception])
+  // FIXME: add this after fixing our own usages
+  // @deprecatedOverriding("Override `def onDownstreamFinish(cause: Throwable)`, instead.", since = "2.6.0") // warns when overriding
```
Added https://github.com/akka/akka/issues/27472 to follow up with this in an extra step. Until then some of our stages don't propagate cancellation causes.

jrudolph commented on akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala:
```diff
@@ -531,10 +532,21 @@ abstract class GraphStageLogic private[stream] (val inCount: Int, val outCount:
+   *
+   * It is recommended to provide a cause.
+   */
+  /* FIXME: Should we deprecate, cancelling without a cause? Seems like a massive change even just for our code.
```
I agree, let's remove that fixme for now and add a comment instead.

jrudolph committed:
Suggest to use `cancel(in, cause)` but don't deprecate old one

jrudolph committed:
Collapse mima exclusions for akka.stream.impl

jrudolph committed:
fix another TODO

jrudolph commented on akka-stream/src/main/scala/akka/stream/impl/fusing/ActorGraphInterpreter.scala:
```diff
@@ -655,7 +660,7 @@ import scala.util.control.NonFatal
       // otherwise this will have no effect
       outputs.foreach(_.fail(reason))
-      inputs.foreach(_.cancel())
+      inputs.foreach(_.cancel(ex)) // TODO: should the reason be wrapped to include information that the stream was aborted?
```
To answer myself: no, it should also use `reason`.

jrudolph committed:
link to FIXME ticket

jrudolph commented:
Addressed all comments and left-over TODOs and FIXMEs. This step is ready from my side. #27472 will do the next steps

johanandren reviewed (approved):
LGTM

One tiny: update MyEventsByTagSource in the docs to overload the new rather than the deprecated onDownstreamFinish

patriknw reviewed (approved):
LGTM
